Block sending existing campaigns for big senders    if sender domain is not authorized

brezocordero opened the pull request:
## Description

Implements the remaining sending rules regarding sender domain authentication.

If using MSS and is a big sender (>1000 subscribers), existing campaigns with unauthenticated domains will be paused during sending.

This rule will be applied to existing users on February 1st.

## Code review notes

I have added a transient to reduce the number of API calls.
The function `checkAuthorizedEmailAddresses` uses `getVerifiedSenderDomainsIgnoringCache` that resets the transient. I believe this function is used in all the places where we would need to update the verified sender domains but please double check I did not forget anything.


## QA notes

We can't test yet the errors from the API but you can test the check before sending:

Reduce the lower_limit and upper_limit so that your subscribers number > upper_limit
Activate a post notification that sends immediately and that uses a verified domain as sender address.
Remove the verified domain from the shop (make sure the address is still authenticated)
Publish a post
Check that the post notification history is not sent and it is paused.

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-5832]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-5832]: https://mailpoet.atlassian.net/browse/MAILPOET-5832?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

brezocordero committed:
Use new function to get verified domains

It uses the overall status set by the shop instead of checking the records.

[MAILPOET-5832]

brezocordero committed:
Add function isAuthorizedDomainRequiredForExistingCampaigns

[MAILPOET-5832]

brezocordero committed:
Modify the check for sender address

to implement rule for existing campaigns.

[MAILPOET-5832]

brezocordero committed:
Pause sending task if sender address not valid

[MAILPOET-5832]

brezocordero committed:
Add new API error for forbidden bulk email

[MAILPOET-5832]

brezocordero committed:
Map ERROR_MESSAGE_BULK_EMAIL_FORBIDDEN

Create a soft error to not block sending.

[MAILPOET-5832]

brezocordero committed:
Process Soft error for domain authorization

Pause the task and do not update counts.

[MAILPOET-5832]

brezocordero committed:
Use a transient to store sender domains

to reduce API calls.

[MAILPOET-5832]

brezocordero committed:
Reduce API calls to verify sender domain during sending

and call checkAuthorizedEmailAddresses on bulk email sending error.

[MAILPOET-5832]

johnolek commented on mailpoet/lib/Services/AuthorizedSenderDomainController.php:
```diff
@@ -288,7 +296,11 @@ public function isSmallSender(): bool {
+    return $this->subscribers->getSubscribersCount() > self::UPPER_LIMIT;
+  }
+
+  private function restrictionsApply() {
```
```suggestion
  private function restrictionsApply(): bool {
```

brezocordero replied:
Done in https://github.com/mailpoet/mailpoet/pull/5391/commits/40f5c52c637c78f71fd013ce58b35c3378e21069

johnolek commented on mailpoet/lib/Services/AuthorizedEmailsController.php:
```diff
@@ -153,16 +153,16 @@ public function onNewsletterSenderAddressUpdate(NewsletterEntity $newsletter, st
     }
 
-    if (!$this->senderDomainController->isAuthorizedDomainRequiredForNewCampaigns()) {
+    $isAuthorizedDomainRequired = $context === 'activation' ?
+      $this->senderDomainController->isAuthorizedDomainRequiredForNewCampaigns() :
+      $this->senderDomainController->isAuthorizedDomainRequiredForExistingCampaigns();
```
Minor style thing, I'm more used to seeing the `?` and `:` of ternaries at the beginning of the lines, and a quick scan through the codebase makes it look like it's much more common.

brezocordero replied:
I can change this :)

brezocordero replied:
Done in https://github.com/mailpoet/mailpoet/pull/5391/commits/a16a6e9d7dfbc28a890cee42d220c9a71f75765d

johnolek commented on mailpoet/lib/Cron/Workers/SendingQueue/SendingQueue.php:
```diff
@@ -208,6 +214,17 @@ private function processSending(ScheduledTaskEntity $task, int $timer): void {
       return;
     }
 
+    // Pause task if sender domain requirements are not met
+    if (!$this->authorizedEmailsController->isSenderAddressValid($newsletter, 'sending')) {
+      $this->loggerFactory->getLogger(LoggerFactory::TOPIC_NEWSLETTERS)->info(
+        'pause task in sending queue due to sender domain requirements',
+        ['task_id' => $task->getId()]
+      );
+      $task->setStatus(ScheduledTaskEntity::STATUS_PAUSED);
+      $this->scheduledTasksRepository->flush();
+      return;
+    }
```
This looks totally fine to me, but I'm curious to know what you think about the placement. We run this check before we start working on batches. Do we ever run into issues where the state changes in between batches and we should pause the sending but don't?

I think this would actually be much more likely to apply to the existing check above looking for deleted/trashed segments, and it's probably not relevant for this code since we're caching results for 24 hours with the transient.

On that note, what does that caching mean for users who have their campaigns paused and then validate their email address? Is there any chance they would have to wait, or can we always rely on [resumeSending](https://github.com/mailpoet/mailpoet/blob/994f46762d141387a06a205b933f5f96557cc887/mailpoet/lib/API/JSON/v1/Mailer.php#L86-L86) to break the cache through [checkAuthorizedEmailAddresses](https://github.com/mailpoet/mailpoet/blob/994f46762d141387a06a205b933f5f96557cc887/mailpoet/lib/Services/AuthorizedEmailsController.php#L125-L125)?

I'm mostly just curious if you think there's an argument for moving checks like these into `enforceSendingAndExecutionLimits` which runs on every batch.

brezocordero replied:
> Do we ever run into issues where the state changes in between batches and we should pause the sending but don't?

I don't think it would be common but it can happen. In that case the API would throw the error and we would pause the task. That's why I added the check for Paused tasks inside of the subscribers loop.
I can add a comment about it.

> On that note, what does that caching mean for users who have their campaigns paused and then validate their email address?

Validating the email address will reset the transient for Sender Domains because it calls `checkAuthorizedEmailAddresses`. 

> I'm mostly just curious if you think there's an argument for moving checks like these into enforceSendingAndExecutionLimits which runs on every batch.

I think it is a good idea and it would have made the work on this ticket much easier. 
I would do it on a separate ticket. We might not need to run all the checks every time `enforceSendingAndExecutionLimits` runs and we would need to be careful with the changes and the performance. This is also a critical part of the application, so taking the time for the refactor and adding tests would be the best.

brezocordero replied:
I added a comment when checking if the task has been paused when processing a batch in https://github.com/mailpoet/mailpoet/pull/5391/commits/c071b7f9b6c26a91b2905be6a42cfe58b7e39442

To move the checks to `enforceSendingAndExecutionLimits`, could you please create a maintenance ticket?

johnolek commented on mailpoet/lib/Cron/Workers/SendingQueue/SendingErrorHandler.php:
```diff
@@ -63,6 +69,14 @@ private function processSoftError(MailerError $error, ScheduledTaskEntity $task,
+      if ($error->getOperation() === MailerError::OPERATION_DOMAIN_AUTHORIZATION) {
+        $this->loggerFactory->getLogger(LoggerFactory::TOPIC_NEWSLETTERS)->info(
+          'Paused task in sending queue due to sender domain authorization error',
+          ['task_id' => $task->getId()]
```
It's probably not even common for people to use multiple sending domains, but for those that do, what do you think about trying to get the actual sender domain or from address in the context? I was thinking it could make debugging easier but I could also see it being redundant if it's clear from other logs or in the UI.

brezocordero replied:
> I could also see it being redundant if it's clear from other logs or in the UI.

Thanks for the comment. There is another log with the sender domain from the function getDmarcMessage. I would skip it adding it here too, for now, and add it later if we find we need it for debugging.

johnolek commented on mailpoet/lib/Services/AuthorizedSenderDomainController.php:
```diff
@@ -235,9 +244,21 @@ private function returnAllDomains(array $records): array {
+    $this->currentRecords = null;
+    $this->currentRawData = $this->bridge->getRawSenderDomainData();
+    $this->wp->setTransient(self::SENDER_DOMAINS_KEY, $this->currentRawData, 60 * 60 * 24);
+  }
```
Extremely minor comment since this looks fine as is.

This method looks slightly inconsistent. It's simply setting `currentRecords` to `null` and doesn't reload them. I think this might be simpler if instead it were called `clearCache`, and you could set `currentRawData` to `null` as well, and also delete the transient.

It would work exactly the same. The only real benefit would be not having to worry about updating the transient time in multiple places, but that could easily be fixed with a class constant... so again, no big deal.

brezocordero replied:
I wanted to avoid another API call. 
I am not fully sure when `$this->currentRecords` would need to be set to `null` or even if we still need it.
I didn't have time to look at it in detail but I can open a Janitorial ticket to do some cleanup here.

brezocordero replied:
I added some cleanup to the ticket to remove the enforcement date, including checking `$this->currentRecords` and the function `reloadCache`

johnolek commented on mailpoet/lib/Mailer/Methods/MailPoet.php:
```diff
@@ -82,14 +82,20 @@ public function send($newsletter, $subscriber, $extraParams = []): array {
+          && !empty($result['error']) && $result['error'] === API::ERROR_MESSAGE_BULK_EMAIL_FORBIDDEN
+        )
+      ) {
+          $this->authorizedEmailsController->checkAuthorizedEmailAddresses();
```
Looks like the indentation is slightly off here

brezocordero replied:
Thanks! The linter missed this but you have the eye.

brezocordero replied:
I used the `switch` instead

johnolek commented on mailpoet/lib/Mailer/Methods/MailPoet.php:
```diff
@@ -82,14 +82,20 @@ public function send($newsletter, $subscriber, $extraParams = []): array {
-      && $result['error'] === API::ERROR_MESSAGE_INVALID_FROM
-    ) {
-      $this->authorizedEmailsController->checkAuthorizedEmailAddresses();
+    if (!empty($result['code'])) {
```
There's something about the formatting of the conditionals that makes this a little hard for me to read. Do you have any guess about if we'll be adding to this in the future? If so (i.e. it might get even more complex) I might personally prefer something more like this (no guarantee that it's functionally the same!):

```php
public function processSendError($result, $subscriber, $newsletter) {
    if (empty($result['code'])) {
        return $this->errorMapper->getErrorForResult($result, $subscriber, $this->sender, $newsletter);
    }

    switch ($result['code']) {
        case API::RESPONSE_CODE_KEY_INVALID:
            $this->bridge->invalidateMssKey();
            break;

        case API::RESPONSE_CODE_CAN_NOT_SEND:
            if ($result['error'] === API::ERROR_MESSAGE_INVALID_FROM) {
                $this->authorizedEmailsController->checkAuthorizedEmailAddresses();
            }
            break;

        case API::RESPONSE_CODE_PAYLOAD_ERROR:
            if (!empty($result['error']) && $result['error'] === API::ERROR_MESSAGE_BULK_EMAIL_FORBIDDEN) {
                $this->authorizedEmailsController->checkAuthorizedEmailAddresses();
            }
            break;
    }

    return $this->errorMapper->getErrorForResult($result, $subscriber, $this->sender, $newsletter);
}
```

brezocordero replied:
Thanks for the suggestion. I will use the `switch`.

brezocordero replied:
Done in https://github.com/mailpoet/mailpoet/pull/5391/commits/0882a36c7bebb612b77833979953aa094038eebd

johnolek reviewed (dismissed):
I left a bunch of comments but none of them seem important, so I'm approving this and letting you decide if anything seems worth changing.

One note I couldn't add as a comment is `getAllSenderDomainsIgnoringCache` doesn't appear to be used anymore so it would be good to remove that method unless we might need it in the future?

I think you could remove the old constants like `DOMAIN_VERIFICATION_STATUS_VALID`, or maybe change the values and use them instead of the ones like `OVERALL_STATUS_VERIFIED`. I don't think the `OVERALL_` distinction matters anymore after these changes but I could be mistaken.

I didn't see anywhere else we would need to reset the transient.

brezocordero committed:
Remove unused code

[MAILPOET-5832]

brezocordero committed:
Add return type

[MAILPOET-5832]

brezocordero committed:
Improve readability

[MAILPOET-5832]

brezocordero committed:
Add comment on check for PAUSED task during batch processing

[MAILPOET-5832]

brezocordero committed:
Improve readability

[MAILPOET-5832]

brezocordero commented:
@johnolek Thanks for the review.

I addressed all the comments here or added them to a ticket.

> One note I couldn't add as a comment is getAllSenderDomainsIgnoringCache doesn't appear to be used anymore so it would be good to remove that method unless we might need it in the future?

I removed it on https://github.com/mailpoet/mailpoet/pull/5391/commits/6ac8e2b6fcfffe2b6ba48407dc8dfcb6da306487

Aschepikov reviewed (approved):
GJ!
